### ISSUE TITLE
Support more codepages via ICU uconv

### ISSUE THREAD

**Issue by roytam1** [RamonUnch/GreenPad#60]
I wonder if we can call ICU's uconv.exe for supporting codepages that windows doesn't support, for example EBCDIC DBCS codepages.

**Comment by RamonUnch** [comment]
This should be possible and would be nice,
I was thinking to dynamically load iconv.dll if it were present. I did not know about uconv, does it have advantages compared to iconv? is there a uconv.dll that would be easier to use than running an exe?

**Comment by roytam1** [comment]
> I did not know about uconv, does it have advantages compared to iconv? is there a uconv.dll that would be easier to use than running an exe?

uconv comes from IBM ICU ( https://icu.unicode.org/ ), and uconv has advantage over iconv is uconv supports EBCDIC DBCS codepages.
and for using EXE instead of DLL is license issues. Running external program doesn't have any license problem. But if you use DLL style then you have to face license issue.
libiconv is GNU GPL license, if you link to it by DLL functions, then your program has to be GPL as well.

**Comment by RamonUnch** [comment]
GPL is quite contagious. 
But would this even apply if the dependency is optional and the dll is not shipped with GreenPad?
There seems to be quite some debates about this so better safe than sorry.

Either way the difference between running an *external optional exe* and an *external optional dll* seems very minor. The difference to me is even more minor than between a *static linking* and non *optional dynamic linking.* which already seems to be a huge debate.

I would suspect that if the optional dll is GPL-contagious, then the optional exe is also contagious, running an external exe otherwise would be an universal trick against GPL contamination.

Either way we could make a very simple conv.dll that would have the same interface than `uconv` but in the public domain, and people would be able to use other dlls.

**Comment by roytam1** [comment]
> Either way the difference between running an _external optional exe_ and an _external optional dll_ seems very minor.

but it involves non-standard API for calling external DLL. Running EXE doesn't involve this.

**Comment by RamonUnch** [comment]
True, plus it would be even more flexible and allow for other converters in the future.

**Comment by RamonUnch** [comment]
I saw that you had made a build of uconv on your blog, nice build, I will experiment with that...

**Comment by roytam1** [comment]
These day I used gemini to write a minimal mb2wc/wc2mb functions with external packed table to convert between multibyte and unicode: https://github.com/roytam1/gmbxwc
Kind of works, but DBCS tables are still quite big (but smaller than NLS files).

Still thinking about how to integrate or bundle as external library.

**Comment by roytam1** [comment]
let LLM helping me to do this:

# Phase 0 results + finalized plan

Probed the real DLL (x64, 20 embedded tables, all display names ASCII as you said). Key semantics, all confirmed by execution:

- **Return** = dest units written; **`dest_max=0` query works** both directions; **truncation stops at char boundaries without setting the flag**.
- **Unmapped flag is sticky-false-unless-set** (init `FALSE`!): stays clear on clean conversions, sets on real loss — `0xFF` byte→`U+FFFD` on read, `U+20000`→`0x3F` on write. (My first probe run showed `un=1` everywhere only because I initialized it `TRUE` — harness bug, not DLL.)
- **Lone trailing lead byte → `U+FFFD` + flag set**: chunk splitting *must* avoid feeding incomplete tails, or every CJK chunk boundary warns falsely.
- **EBCDIC statefulness inconclusive** (my hand-made IBM930 bytes were invalid — all paths returned junk differently). Design below is robust under both possible semantics (details in §3).

## Implementation plan

**0. Arch DLL names** — `GMBXWC_DLL` block mirroring chardet's `#if defined(_M_…)` chain: default `gmbxwc.dll`, plus `gmbxwc_x64/_ia64/_arm64/_arm/_mips/_ppc/_axp64/_axp.dll`. Assumption flagged: exact tokens are one `#define` block, trivial to rename when you confirm real filenames.

**1. Loader (`kilib/textfile.cpp` only)** — chardet pattern: exe-relative existence check (Win32s-safe) → `LoadLibrary` → `GmbxwcApi_Load` (all 10) → cached handle+table for process lifetime (unlike chardet's per-call load; enumeration + every file op would thrash otherwise). Single-threaded app, no races; contexts freed per file so the header's free-before-unload rule holds vacuously. `NO_GMBXWC` build flag mirroring `NO_CHARDET`.

**2. Reader `rGmbxwc`** (mirrors `rMBCS` structure): per-session `CodePageContext*` (`CreateConverter` at construction, `Free…` in dtor) + **unconditional ≤7-byte carry-over** at non-final chunk ends (covers ≤8-byte chars incl. GB18030; ~0.005% reconvert cost; table-agnostic) + **SI/SO mode tracking** (`0x0E/0x0F` scan, back off cuts to the last SI when inside DBCS mode). The SI back-off is correct whether the DLL keeps shift state across calls or restarts in SBCS mode; only pathological >chunk-size DBCS runs degrade (documented). Unmapped input → same replacement policy as existing readers.

**3. Writer `wGmbxwc`**: same per-session context; sizing via confirmed `dest_max=0` query with grow-and-retry fallback; lossy chars → table `def_char` (`0x3F` observed).

**4. Routing**: `if (30000 <= cs && cs < 40000)` range branches around both `TextFileR/W::Open` switches; validate `index < EmbeddedCount`, DLL present — else `Open` fails into the existing open-error path. `codepage()` already returns the raw ID.

**5. Enrollment** (`CharSetList` ctor, after static entries, gated on load success — absent DLL = zero entries, fully silent): `ID = 30000+i`, `longName = display_name`, `shortName = "CP<codepage>"` in `CharSetList`-owned buffers. Type `BOTH` → open/save/reopen/config combos + status bar work with no dialog-code changes.

**6. Consumers** (verify-only): cmdline `-c30001` flows as int already; INI ints; `MayLoseData`→true + `HasLossyChars`/`FindLossyChar` branch via `IndexedWC2MB` unmapped flag (else every gmbxwc save would warn spuriously).

**7. Tests**: round-trip all 20 tables; unmapped both directions; chunk-split incl. synthetic EBCDIC SI/SO stream; missing-DLL paths; enrollment count; `-c30001`; save-warning; plus syntax/cp932/build discipline. Docs: `readme.en.txt` charset list.

and pushed as https://github.com/roytam1/GreenPad/commit/248a61c4fcacc28980c277ae6c70afb6e65ff9e6

preq: LLM helped me to better handling of lossy conversion to multibyte:
https://github.com/roytam1/GreenPad/commit/336127ca2e0a41db46746fd3024a62e0025242de
https://github.com/roytam1/GreenPad/commit/a64ae38e2868d4e5ffc332d8343fbf4500085ff8
